### ISSUE TITLE
Function on main thread

### ISSUE THREAD

**Issue by bkj** [channelcat/sanic#653]
This may be a bit from left field -- but I'll ask anyway.

I'm trying to run a neural network behind a REST api.  The model
  a) grabs an image from the web
  b) runs neural network on the GPU

Because the model runs on the GPU, it's not playing nicely with multiple threads.  The IO is much slower than the model, so I think a solution would be to have multiple workers handling loading the images and then pass them to a single thread running the model.  Any thoughts on how that might be accomplished?

Here's roughly what I have right now

```
model = load_neural_network()

@app.route('/api/score', methods=['POST'])
async def post_handler(request):
    url = request.json['url']
    img = url2image(url)
    pred = model.predict(img)
    return response.json({'response' : pred)})

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=9000, workers=1)
```

This works, but if I increase the number of `workers` it fails.  I think there may be some parallel between this and using a global connection pool as in 

https://github.com/channelcat/sanic/blob/master/examples/sanic_aiomysql_with_global_pool.py

but I'm not completely sure.

**Comment by r0fls** [comment]
What error are you receiving, or how does it fail?

**Comment by bkj** [comment]
Usually I get an OOM error from the GPU -- it appears that the server tries to load the model several times simultaneously which takes up all the GPU RAM and eventually throws an error.

**Comment by seemethere** [comment]
@bkj Yeah that's a consequence of multiprocessing and the way Python actually handles multiple processes where it just takes a copy of the current running program and multiplies it over `N` amount of processes.

If you want you could just run it single worker and you wouldn't experience these issues. What could work too is if you had a single API running the GPU process which communicates back and forth with your outward facing API.

Not entirely sure how fast it would be but it'd be worth a shot.

**Comment by bkj** [comment]
Yeah I ended up implementing a version of the latter.  It can be significantly faster, depending on the network latency, GPU utilization etc.  Somewhat off topic -- but if you were to implement that kind of thing, how would you do the communication between the processes?  I did it using Redis, but was trying to figure out if there's a better way (eg to avoid serializing `numpy` arrays, etc)

**Comment by seemethere** [comment]
I'm always a fan of implementing it through a REST API because that's the type of interface that I'm used to and a lot of people are used to but Redis is also a good choice.

As far as not having to serialize the numpy arrays it's gonna be tough if it's not running in the same process. Maybe it would be better more performance effective to just serialize the parts of the numpy array that you actually need instead of the whole thing?
